crypto/keys/multisig/multisig.go

```diff
@@ -0,0 +1,128 @@
+	pubKeys := m.GetPubKeys()
+	// ensure bit array is the correct size
+	if len(pubKeys) != size {
+		return fmt.Errorf("bit array size is incorrect %d", len(pubKeys))
+	}
+	// ensure size of signature list
+	if len(sigs) < int(m.Threshold) || len(sigs) > size {
+		return fmt.Errorf("signature size is incorrect %d", len(sigs))
+	}
+	// ensure at least k signatures are set
+	if bitarray.NumTrueBitsBefore(size) < int(m.Threshold) {
+		return fmt.Errorf("minimum number of signatures not set, have %d, expected %d", bitarray.NumTrueBitsBefore(size), int(m.Threshold))
+	}
+	// index in the list of signatures which we are concerned with.
+	sigIndex := 0
+	for i := 0; i < size; i++ {
+		if bitarray.GetIndex(i) {
+			si := sig.Signatures[sigIndex]
+			switch si := si.(type) {
+			case *signing.SingleSignatureData:
+				msg, err := getSignBytes(si.SignMode)
+				if err != nil {
+					return err
+				}
+				if !pubKeys[i].VerifySignature(msg, si.Signature) {
+					return err
+				}
+			case *signing.MultiSignatureData:
+				nestedMultisigPk, ok := pubKeys[i].(multisig.PubKey)
+				if !ok {
+					return fmt.Errorf("unable to parse pubkey of index %d", i)
+				}
+				if err := nestedMultisigPk.VerifyMultisignature(getSignBytes, si); err != nil {
+					return err
+				}
+			default:
+				return fmt.Errorf("improper signature data type for index %d", sigIndex)
+			}
+			sigIndex++
+		}
+	}
+	return nil
+}
+
+// VerifySignature implements crypto.PubKey VerifySignature method,
+// it panics because it can't handle MultiSignatureData
+// cf. https://github.com/cosmos/cosmos-sdk/issues/7109#issuecomment-686329936
+func (m *LegacyAminoMultisigThresholdPubKey) VerifySignature(msg []byte, sig []byte) bool {
+	panic("not implemented")
+}
+
+// GetPubKeys implements the PubKey.GetPubKeys method
+func (m *LegacyAminoMultisigThresholdPubKey) GetPubKeys() []tmcrypto.PubKey {
+	if m != nil {
+		pubKeys := make([]tmcrypto.PubKey, len(m.PubKeys))
+		for i := 0; i < len(m.PubKeys); i++ {
+			pubKeys[i] = m.PubKeys[i].GetCachedValue().(tmcrypto.PubKey)
+		}
+		return pubKeys
+	}
+
+	return nil
+}
+
+// Equals returns true if m and other both have the same number of keys, and
+// all constituent keys are the same, and in the same order.
+func (m *LegacyAminoMultisigThresholdPubKey) Equals(key tmcrypto.PubKey) bool {
+	otherKey, ok := key.(multisig.PubKey)
+	if !ok {
+		return false
+	}
+	pubKeys := m.GetPubKeys()
+	otherPubKeys := otherKey.GetPubKeys()
+	if m.GetThreshold() != otherKey.GetThreshold() || len(pubKeys) != len(otherPubKeys) {
+		return false
+	}
+
+	for i := 0; i < len(pubKeys); i++ {
+		if !pubKeys[i].Equals(otherPubKeys[i]) {
+			return false
+		}
+	}
+	return true
+}
+
+// GetThreshold implements the PubKey.GetThreshold method
+func (m *LegacyAminoMultisigThresholdPubKey) GetThreshold() uint {
+	return uint(m.Threshold)
+}
+
+// Type returns multisig type
+func (m *LegacyAminoMultisigThresholdPubKey) Type() string {
+	return "PubKeyMultisigThreshold"
```

Comment:
Should this be `LegacyAminoPubKeyMultisigThreshold `?

Reply:
Makes sense. I also added the `cosmos/` prefix in 4114ce9a001619289ee473b01b34de57e498032a, not sure which convention to take here.

Reply:
Ah wait, no. This changes the bech32 formatting, so we must keep the same names as before. I also kept `tendermint/{Pub,Priv}KeySecp256k1` for the secp256k1 keys.